ldap_auth_provider.py

```diff
@@ -220,6 +210,105 @@ def check_password(self, user_id, password):
             logger.warning("Error during ldap authentication: %s", e)
             defer.returnValue(False)
 
+    @defer.inlineCallbacks
+    def check_3pid_auth(self, medium, address, password):
+        """ Handle authentication against thirdparty login types, such as email
+
+            Args:
+                medium (str): Medium of the 3PID (e.g email, msisdn).
+                address (str): Address of the 3PID (e.g bob@example.com for email).
+                password (str): The provided password of the user.
+
+            Returns:
+                user_id (str|None): ID of the user if authentication
+                    successful. None otherwise.
+        """
+
+        # We currently only support email
+        if medium != "email":
+            return
+
+        # Talk to LDAP and check if this email/password combo is correct
+        try:
+            server = ldap3.Server(self.ldap_uri, get_info=None)
+            logger.debug(
+                "Attempting LDAP connection with %s",
+                self.ldap_uri
+            )
+
+            search_filter = [("mail", address)]
+            result, conn, response = yield self._ldap_authenticated_search(
+                server=server, password=password, filters=search_filter,
+                attributes=['givenname', 'cn'],
+            )
+            logger.debug(
+                'LDAP auth method authenticated search returned: '
+                '%s (conn: %s) (response: %s)',
+                result,
+                conn,
+                response
+            )
+            if not result:
+                defer.returnValue(None)
+
+            try:
+                logger.info(
+                    "User authenticated against LDAP server: %s",
+                    conn
+                )
+            except NameError:
+                logger.warning(
+                    "Authentication method yielded no LDAP connection, "
+                    "aborting!"
+                )
+                defer.returnValue(None)
+
+            # Extract the username from the search response from the LDAP server
+            localpart = response["attributes"].get("cn", [""])[0]
+            givenName = response["attributes"].get("givenName", [localpart])[0]
+
+            # Register the user
+            user_id = yield self.register_user(localpart, givenName, address)
+
+            defer.returnValue(user_id)
+
+        except ldap3.core.exceptions.LDAPException as e:
+            logger.warning("Error during ldap authentication: %s", e)
+            defer.returnValue(None)
```

Comment:
I'd have thought we wouldn't want to silently continue, as that is going to result in the wrong thing?

Reply:
Again taken from elsewhere in the file. I agree silently failing isn't great for a client's perspective (they just get "invalid username/password"). Alternative is to actually fail so a M_UNKNOWN is returned which may actually prompt someone to poke their server admin.

Our only options for password providers are to return a `user_id` or `None`, either meaning authentication success or fail. We could just `raise` here though?

Reply:
Hmm, ok. `raise` was what I vote for but we should be consistent.

Reply:
So do I. I'll edit all occurrences.